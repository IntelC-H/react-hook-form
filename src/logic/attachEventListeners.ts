import { Field } from '../type';
import isCheckbox from '../utils/isCheckBox';

export default function attachEventListeners({
  field,
  validateAndStateUpdate,
  isRadio,
}: {
  field: Field;
  isRadio: boolean;
  validateAndStateUpdate: (any) => void;
}): void {
  const { ref } = field;
  ref.addEventListener(isCheckbox(ref.type) || isRadio ? 'change' : 'input', validateAndStateUpdate);
  ref.addEventListener('blur', validateAndStateUpdate);
}
