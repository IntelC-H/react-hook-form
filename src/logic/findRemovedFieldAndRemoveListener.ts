import removeAllEventListeners from './removeAllEventListeners';
import isRadioInput from '../utils/isRadioInput';
import isCheckBoxInput from '../utils/isCheckBoxInput';
import isDetached from '../utils/isDetached';
import isArray from '../utils/isArray';
import { Field, FieldRefs, FieldValues } from '../types';

export default function findRemovedFieldAndRemoveListener<
  FormValues extends FieldValues
>(
  fields: FieldRefs<FormValues>,
  handleChange: ({ type, target }: Event) => Promise<void | boolean>,
  field: Field,
  forceDelete?: boolean,
): void {
  if (!field) {
    return;
  }

  const {
    ref,
    ref: { name, type },
    mutationWatcher,
  } = field;

  if (!type) {
    delete fields[name];
    return;
  }

  const fieldValue = fields[name];

  if ((isRadioInput(ref) || isCheckBoxInput(ref)) && fieldValue) {
    const { options } = fieldValue;

    if (isArray(options) && options.length) {
      options.forEach(({ ref, mutationWatcher }, index): void => {
        if ((ref && isDetached(ref)) || forceDelete) {
          removeAllEventListeners(ref, handleChange);

          if (mutationWatcher) {
            mutationWatcher.disconnect();
          }

          options.splice(index, 1);
        }
      });

      if (options && !options.length) {
        delete fields[name];
      }
    } else {
      delete fields[name];
    }
  } else if (isDetached(ref) || forceDelete) {
    removeAllEventListeners(ref, handleChange);

    if (mutationWatcher) {
      mutationWatcher.disconnect();
    }
    delete fields[name];
  }
}
