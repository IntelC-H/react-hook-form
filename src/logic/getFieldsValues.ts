import getFieldValue from './getFieldValue';
import isString from '../utils/isString';
import isArray from '../utils/isArray';
import isUndefined from '../utils/isUndefined';
import { InternalFieldName, FieldValues, FieldRefs } from '../types/types';

export default <TFieldValues extends FieldValues>(
  fields: FieldRefs<TFieldValues>,
  search?:
    | InternalFieldName<TFieldValues>
    | InternalFieldName<TFieldValues>[]
    | { nest: boolean },
) => {
  const output = {} as TFieldValues;

  for (const name in fields) {
    if (
      isUndefined(search) ||
      (isString(search)
        ? name.startsWith(search)
        : isArray(search) && search.find((data) => name.startsWith(data)))
    ) {
      output[name as InternalFieldName<TFieldValues>] = getFieldValue(
        fields,
        fields[name]!.ref,
      );
    }
  }

  return output;
};
