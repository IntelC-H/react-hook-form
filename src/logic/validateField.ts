import getRadioValue from './getRadioValue';
import isRadioInput from '../utils/isRadioInput';
import { DATE_INPUTS, STRING_INPUTS } from '../constants';
import { ErrorMessages, Field } from '..';
import getValueAndMessage from './getValueAndMessage';

export default (
  { ref: { type, value, name, checked }, options, required, maxLength, minLength, min, max, pattern, validate }: Field,
  fields: { [key: string]: Field },
): ErrorMessages => {
  const copy = {};

  if (
    required &&
    ((type === 'checkbox' && !checked) ||
      (!isRadioInput(type) && type !== 'checkbox' && value === '') ||
      (isRadioInput(type) && !getRadioValue(fields[name].options).isValid))
  ) {
    copy[name] = {
      required,
    };
  }

  if (value === '') return copy;

  // min and max section
  if (min || max) {
    let exceedMax;
    let exceedMin;
    const valueNumber = parseFloat(value);

    const { value: maxValue, message: maxMessage } = getValueAndMessage(max);
    const { value: minValue, message: minMessage } = getValueAndMessage(min);

    if (type === 'number') {
      exceedMax = maxValue && valueNumber > maxValue;
      exceedMin = minValue && valueNumber < minValue;
    } else if (DATE_INPUTS.includes(type)) {
      if (typeof maxValue === 'string') exceedMax = maxValue && new Date(value) > new Date(maxValue);
      if (typeof minValue === 'string') exceedMin = minValue && new Date(value) < new Date(minValue);
    }

    if (exceedMax || exceedMin) {
      copy[name] = {
        ...copy[name],
        ...(exceedMax ? { max: maxMessage } : null),
        ...(exceedMin ? { min: minMessage } : null),
      };
    }
  }

  if ((maxLength || minLength) && STRING_INPUTS.includes(type)) {
    const { value: maxLengthValue, message: maxLengthMessage } = getValueAndMessage(maxLength);
    const { value: minLengthValue, message: minLengthMessage } = getValueAndMessage(minLength);

    const exceedMax = maxLength && value.toString().length > maxLengthValue;
    const exceedMin = minLength && value.toString().length < minLengthValue;

    if (exceedMax || exceedMin) {
      copy[name] = {
        ...copy[name],
        ...(exceedMax ? { maxLength: maxLengthMessage } : null),
        ...(exceedMin ? { minLength: minLengthMessage } : null),
      };
    }
  }

  if (pattern) {
    const { value: patternValue, message: patternMessage } = getValueAndMessage(pattern);
    if (patternValue instanceof RegExp && !patternValue.test(value)) {
      copy[name] = {
        ...copy[name],
        pattern: patternMessage,
      };
    }
  }

  if (validate) {
    const fieldValue = isRadioInput(type) ? getRadioValue(options).value : value;

    if (typeof validate === 'function') {
      const result = validate(fieldValue);
      if (typeof result !== 'boolean' || !result) {
        copy[name] = {
          ...copy[name],
          validate: result || true,
        };
      }
    } else if (typeof validate === 'object') {
      const result = Object.entries(validate).reduce((previous, [key, validate]) => {
        const result = typeof validate === 'function' && validate(fieldValue);
        if (typeof result !== 'boolean' || !result) {
          previous[key] = result || true;
        }
        return previous;
      }, {});

      if (Object.keys(result).length) {
        copy[name] = {
          ...copy[name],
          validate: result,
        };
      }
    }
  }

  return copy;
};
