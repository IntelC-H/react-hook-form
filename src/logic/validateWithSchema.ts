import { DataType, FieldErrors, ValidationReturn } from '../types';

export function parseErrorSchema(error: DataType): FieldErrors {
  return error.inner.reduce(
    (previous: DataType, { path, message, type }: DataType): FieldErrors => ({
      ...previous,
      [path]: { message, ref: {}, type },
    }),
    {},
  );
}

export default async function validateWithSchema(
  ValidationSchema: any,
  data: DataType,
): Promise<ValidationReturn> {
  try {
    return {
      result: await ValidationSchema.validate(data, { abortEarly: false }),
      fieldErrors: {},
    };
  } catch (e) {
    return {
      fieldErrors: parseErrorSchema(e),
      result: {},
    };
  }
}
