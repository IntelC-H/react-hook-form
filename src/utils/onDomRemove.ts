import { Ref } from '../type';

export default function onDomRemove(element: Ref, onDetachCallback: () => void): any {
  const observer = new MutationObserver(() => {
    function isDetached(el): boolean {
      if (el.parentNode === window.document) {
        return false;
      } else if (el.parentNode === null) {
        return true;
      }

      return isDetached(el.parentNode);
    }

    if (isDetached(element)) {
      observer.disconnect();
      onDetachCallback();
    }
  });

  observer.observe(window.document, {
    childList: true,
    subtree: true,
  });

  return observer;
}
