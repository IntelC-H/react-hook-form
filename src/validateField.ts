import getValidRadioValue from './getValidRadioValue';
import { RegisterInput } from './index';

export default ({ ref, required, maxLength, min, max, pattern }: RegisterInput, fields, errorsRef = {}) => {
  const copy = { ...errorsRef };

  if (
    (ref.type === 'checkbox' && required && !ref.checked) ||
    (ref.type === 'select-one' && required && ref.value === '') ||
    (ref.type === 'textarea' && required && ref.value === '') ||
    (ref.type === 'radio' && required && !getValidRadioValue(fields, ref.name).isValid) ||
    (ref.type === 'number' && required && (ref.value === 0 || ref.value === '')) ||
    ((ref.type === 'text' || ref.type === 'password' || ref.type === 'file' || ref.type === 'image') &&
      required &&
      ref.value === '')
  ) {
    copy[ref.name] = {
      required: true,
    };
  }

  // min and max section
  if (((min || max) && ref.type === 'number' && (max && ref.value > max)) || (min && ref.value < min)) {
    copy[ref.name] = {
      ...copy[ref.name],
      ...(ref.value > max ? { max: true } : null),
      ...(ref.value < min ? { min: true } : null),
    };
  }

  if (maxLength && ref.type === 'text' && maxLength && typeof ref.value === 'string' && ref.value.length > max) {
    copy[ref.name] = {
      ...copy[ref.name],
      maxLength: true,
    };
  }

  if (
    pattern &&
    ref.type === 'text' &&
    typeof ref.value === 'string' &&
    pattern instanceof RegExp &&
    !pattern.test(ref.value)
  ) {
    copy[ref.name] = {
      ...copy[ref.name],
      pattern: true,
    };
  }

  return copy;
};
