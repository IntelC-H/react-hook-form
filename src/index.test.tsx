import * as React from 'react';
import useForm from './';
import { act } from 'react-dom/test-utils';
import attachEventListeners from './logic/attachEventListeners';
import getFieldsValues from './logic/getFieldsValues';
import findRemovedFieldAndRemoveListener from './logic/findRemovedFieldAndRemoveListener';
import validateWithSchema from './logic/validateWithSchema';
import validateField from './logic/validateField';
import onDomRemove from './utils/onDomRemove';
import { mount } from 'enzyme';

jest.mock('./utils/onDomRemove');
jest.mock('./logic/findRemovedFieldAndRemoveListener');
jest.mock('./logic/validateField');
jest.mock('./logic/attachEventListeners');
jest.mock('./logic/getFieldsValues');
jest.mock('./logic/validateWithSchema');

let hookForm;
let hookFormWithValidationSchema;
let wrapper;

const testComponent = callback => {
  const TestHook = ({ callback }) => {
    const { errors } = callback();
    return errors ? <div>errors</div> : null;
  };

  wrapper = mount(<TestHook callback={callback} />);
};

describe('useForm', () => {
  beforeEach(() => {
    testComponent(() => {
      hookForm = useForm();
      return hookForm;
    });

    jest.resetAllMocks();
  });

  describe('register', () => {
    it('should return undefined when ref is undefined', () => {
      expect(hookForm.register(undefined)).toBeUndefined();
    });

    it('should return undefined when ref name is missing', () => {
      expect(hookForm.register({ type: 'input' }, {})).toBeUndefined();
    });

    it('should register field and call attachEventListeners method', () => {
      hookForm.register({ type: 'input', name: 'test' });
      expect(attachEventListeners).toBeCalledWith({
        field: {
          mutationWatcher: undefined,
          ref: {
            name: 'test',
            type: 'input',
          },
        },
        isRadio: false,
        validateAndStateUpdate: expect.any(Function),
      });
      expect(onDomRemove).toBeCalled();
    });

    it('should register field for radio type and call attachEventListeners method', () => {
      hookForm.register({ type: 'radio', name: 'test' });
      expect(attachEventListeners).toBeCalledWith({
        field: {
          mutationWatcher: undefined,
          ref: {
            name: 'test',
            type: 'radio',
          },
        },
        isRadio: true,
        validateAndStateUpdate: expect.any(Function),
      });
      expect(onDomRemove).toBeCalled();
    });
  });

  describe('watch', () => {
    it('should watch individual input', () => {
      expect(hookForm.watch('test')).toBeUndefined();
      // @ts-ignore
      getFieldsValues.mockImplementation((_, name) => {
        if (name === 'test') {
          return 'data';
        }
      });
      expect(hookForm.watch('test')).toBe('data');
    });

    it('should watch array of inputs', () => {
      expect(hookForm.watch(['test', 'test1'])).toBeUndefined();
      // @ts-ignore
      getFieldsValues.mockImplementation((_, name) => {
        if (name && name.includes('test1') && name.includes('test')) {
          return ['data1', 'data2'];
        }
      });
      expect(hookForm.watch(['test', 'test1'])).toEqual(['data1', 'data2']);
    });

    it('should watch every fields', () => {
      expect(hookForm.watch()).toBeUndefined();
      // @ts-ignore
      getFieldsValues.mockImplementation((_, name) => {
        if (name === undefined) {
          return ['data1', 'data2'];
        }
      });
      expect(hookForm.watch()).toEqual(['data1', 'data2']);
    });
  });

  describe('reset', () => {
    it('should reset the form and re-render the form', () => {
      hookForm.register({ name: 'test' });
      hookForm.setValue('test', 'data');
      hookForm.reset();
      expect(findRemovedFieldAndRemoveListener).toBeCalled();
    });
  });

  describe('unSubscribe', () => {
    it('should remove all reference when mode change', () => {
      hookForm.register({ type: 'input', name: 'test' });
      hookForm.register({
        type: 'radio',
        name: 'test1',
        options: [
          { type: 'radio', name: 'test3' },
          { type: 'radio', name: 'test4' },
        ],
      });
      expect(attachEventListeners).toBeCalledWith({
        field: {
          mutationWatcher: undefined,
          ref: {
            name: 'test',
            type: 'input',
          },
        },
        isRadio: false,
        validateAndStateUpdate: expect.any(Function),
      });
      hookForm.register({ type: 'input', name: 'test' });
      act(() => {
        hookForm.unSubscribe();
      });
      expect(findRemovedFieldAndRemoveListener).toBeCalled();
      hookForm.register({ type: 'input', name: 'test' });
      expect(attachEventListeners).toBeCalledTimes(3);
    });
  });

  describe('handleSubmit', () => {
    it('should invoke the callback when validation pass', async () => {
      const callback = jest.fn();
      const preventDefault = jest.fn();
      const persist = jest.fn();
      await hookForm.handleSubmit(callback)({
        preventDefault,
        persist,
      });
      expect(callback).toBeCalled();
    });

    it('should not invoke callback when there are errors', async () => {
      hookForm.register(
        { value: '', type: 'input', name: 'test' },
        { required: true },
      );
      const callback = jest.fn();
      // @ts-ignore
      validateField.mockImplementation(async () => {
        return { test: { type: 'test' } };
      });
      await hookForm.handleSubmit(callback)({
        preventDefault: () => {},
        persist: () => {},
      });
      expect(callback).not.toBeCalled();
    });

    it('should only validate fields which have been specified', async () => {
      const callback = jest.fn();
      testComponent(() => {
        hookForm = useForm({
          mode: 'onSubmit',
          validationFields: ['test'],
        });
        return hookForm;
      });
      hookForm.register(
        { value: '', type: 'input', name: 'test1' },
        { required: true },
      );
      hookForm.register({ value: '', type: 'input', name: 'test' });
      // @ts-ignore
      validateField.mockImplementation(async () => {
        return {};
      });
      await hookForm.handleSubmit(callback)();
      expect(validateField).toHaveBeenCalledTimes(1);
    });
  });

  describe('getValues', () => {
    it('should call getFieldsValues and return all values', () => {
      hookForm.register({ value: 'test', type: 'input', name: 'test' });
      hookForm.getValues();
      expect(getFieldsValues).toBeCalled();
    });
  });

  describe('handleSubmit with validationSchema', () => {
    it('should invoke callback when error not found', async () => {
      testComponent(() => {
        hookFormWithValidationSchema = useForm({
          mode: 'onSubmit',
          validationSchema: {},
        });
        return hookFormWithValidationSchema;
      });

      hookFormWithValidationSchema.register(
        { value: '', type: 'input', name: 'test' },
        { required: true },
      );
      const callback = jest.fn();
      // @ts-ignore
      getFieldsValues.mockImplementation(async () => {
        return { test: 'test' };
      });
      // @ts-ignore
      validateWithSchema.mockImplementation(async () => {
        return {};
      });
      await hookFormWithValidationSchema.handleSubmit(callback)({
        preventDefault: () => {},
        persist: () => {},
      });
      expect(callback).toBeCalled();
    });
  });

  describe('when component unMount', () => {
    it('should call unSubscribe', () => {
      hookForm.register({ type: 'text', name: 'test' });
      wrapper.unmount();
      expect(findRemovedFieldAndRemoveListener).toBeCalled();
    });
  });

  describe('when defaultValues is supplied', () => {
    it('should populate the input with them', async () => {
      testComponent(() => {
        hookForm = useForm({
          mode: 'onSubmit',
          defaultValues: {
            test: 'data',
          },
        });
        return hookForm;
      });
      // @ts-ignore
      validateField.mockImplementation(async () => {
        return {};
      });
      hookForm.register({ type: 'text', name: 'test' });
      await hookForm.handleSubmit(data => {
        expect(data).toEqual({
          test: 'data',
        });
      })({
        preventDefault: () => {},
        persist: () => {},
      });
    });
  });
});
