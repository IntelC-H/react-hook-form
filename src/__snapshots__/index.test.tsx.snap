// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useForm should handle submit when there is no error 1`] = `ReactWrapper {}`;
